ironfish/src/genesis/addGenesisTransaction.ts

```diff
@@ -0,0 +1,131 @@
+  Asset,
+  Note as NativeNote,
+  Transaction as NativeTransaction,
+} from '@ironfish/rust-nodejs'
+import { Logger } from '../logger'
+import { IronfishNode } from '../node'
+import { Block, BlockHeader } from '../primitives'
+import { transactionCommitment } from '../primitives/blockheader'
+import { Transaction } from '../primitives/transaction'
+import { CurrencyUtils } from '../utils'
+import { Account } from '../wallet'
+import { GenesisBlockAllocation } from './makeGenesisBlock'
+
+export async function addGenesisTransaction(
+  node: IronfishNode,
+  account: Account,
+  allocations: GenesisBlockAllocation[],
+  logger: Logger,
+): Promise<{ block: Block }> {
+  logger = logger.withTag('addGenesisTransaction')
+
+  if (!account.spendingKey) {
+    throw new Error('Must be a full account, not a view account')
+  }
+
+  // Sum the allocations to get the total number of coins
+  const allocationSum = allocations.reduce((sum, cur) => sum + cur.amountInOre, 0n)
+  const allocationSumInIron = CurrencyUtils.encodeIron(allocationSum)
+
+  logger.info('Generating a transaction for distributing allocations...')
+
+  // Get a previous note owned by the given account from the existing genesis block
+  let note: NativeNote | null = null
+  let witness = null
+  const genesisTransactions = await node.chain.getBlockTransactions(node.chain.genesis)
+  for (const { transaction, initialNoteIndex } of genesisTransactions) {
+    let noteIndex = -1
+    for (const encryptedNote of transaction.notes) {
+      noteIndex += 1
+      const decryptedNote = encryptedNote.decryptNoteForOwner(account.incomingViewKey)
+      if (decryptedNote == null) {
+        continue
+      }
+
+      witness = await node.chain.notes.witness(initialNoteIndex + noteIndex)
+      note = decryptedNote.takeReference()
+      decryptedNote.returnReference()
+      break
+    }
+
+    if (note != null) {
+      break
+    }
+  }
+
+  if (note == null) {
+    throw new Error(
+      'The given account does not have a suitable note to spend for the new allocations',
+    )
+  }
+
+  if (witness == null) {
+    throw new Error('The witness is missing, this should not happen')
+  }
+
+  if (note.value() !== allocationSum) {
+    throw new Error('The value of the note to spend does not match the sum of the allocations')
+  }
+
+  // Create the new transaction to be appended to the new genesis block
+  const transaction = new NativeTransaction(account.spendingKey)
+  logger.info(`  Generating a spend for ${allocationSumInIron} coins...`)
+  transaction.spend(note, witness)
+
+  for (const alloc of allocations) {
+    logger.info(
+      `  Generating an output for ${CurrencyUtils.encodeIron(alloc.amountInOre)} coins for ${
+        alloc.publicAddress
+      }...`,
+    )
+    const note = new NativeNote(
+      alloc.publicAddress,
+      BigInt(alloc.amountInOre),
+      alloc.memo,
+      Asset.nativeId(),
+      account.publicAddress, // TODO: Should this be a dummy address instead?
```

Comment:
I think it's fine for it to come from the originating account -- this is a pretty specific usecase anyway